src/index/Index.h

```diff
@@ -599,4 +599,82 @@ class Index {
     }
     LOG(INFO) << "Done" << std::endl;
   }
+
+  // _____________________________________________________________________________
+  template <class PermutationInfo, class MetaData>
+  void scan(const string& keyFirst, const string& keySecond, IdTable* result,
+            const PermutationInfo& perm, const MetaData& meta,
+            ad_utility::File& file) const {
+    LOG(DEBUG) << "Performing " << perm._readableName << "  scan of relation "
+               << keyFirst << " with fixed subject: " << keySecond << "...\n";
+    Id relId;
+    Id subjId;
+    if (_vocab.getId(keyFirst, &relId) && _vocab.getId(keySecond, &subjId)) {
+      if (meta.relationExists(relId)) {
+        auto rmd = meta.getRmd(relId);
+        if (rmd.hasBlocks()) {
+          pair<off_t, size_t> blockOff =
+              rmd._rmdBlocks->getBlockStartAndNofBytesForLhs(subjId);
+          // Functional relations have blocks point into the pair index,
+          // non-functional relations have them point into lhs lists
+          if (rmd.isFunctional()) {
+            scanFunctionalRelation(blockOff, subjId, file, result);
+          } else {
+            pair<off_t, size_t> block2 =
+                rmd._rmdBlocks->getFollowBlockForLhs(subjId);
+            scanNonFunctionalRelation(blockOff, block2, subjId, file,
+                                      rmd._rmdBlocks->_offsetAfter, result);
+          }
+        } else {
+          // If we don't have blocks, scan the whole relation and filter /
+          // restrict.
+          IdTable fullRelation(2);
+          fullRelation.resize(rmd.getNofElements());
+          file.read(fullRelation.data(), rmd.getNofElements() * 2 * sizeof(Id),
+                    rmd._rmdPairs._startFullIndex);
+          getRhsForSingleLhs(fullRelation, subjId, result);
+        }
+      } else {
+        LOG(DEBUG) << "No such relation.\n";
+      }
+    } else {
+      LOG(DEBUG) << "No such second order key.\n";
+    }
+    LOG(DEBUG) << "Scan done, got " << result->size() << " elements.\n";
+  }
+
+  //_____________________________________________________________________________
+  template <class MetaData>
+  void scanImpl(Id key, IdTable* result, const MetaData& meta,
```

Comment:
I think this should have a name that better distinguishes it from the `scan()` function. It's not really an `…Impl` as it isn't used to implement `scan()` right? Maybe `scanFull()` because it scans a full "relation"?

Reply:
I have a different suggestion: Everything is a `scan`. The overload is sufficient to disambiguate:

- 1 or 2 keys
- are the keys strings (have to be looked up in vocab) or Ids (this has already been done)